### ISSUE TITLE
Ctrl+C should stop running TabPy

### ISSUE THREAD

**Issue by 0golovatyi** [tableau/TabPy#345]
Ctrl+C combination doesn't work and Ctrl+Break should be used instead. The second one is not easy to use on some platforms and/or VMs.

Need to understand why Ctrl+C doesn't work, and either fix the behavior or document it.

**Comment by lotus42** [comment]
I'm curious if one of the answers provided here might give some insight on why it doesn't work:
https://stackoverflow.com/questions/1364173/stopping-python-using-ctrlc

I don't think it's a try/except issue from looking at the source, unless I missed something, but maybe something else is blocking it? I suppose if repeatedly hitting Ctrl+C eventually works- then that's the case.

**Comment by siqueira-ec** [comment]
Unless I missed something and based on [Stopping python using ctrl+c](https://stackoverflow.com/a/1364409), I think that we should put main() execution on tabpy module on a try...except block.

Like this:

```python
if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print("ctrl+c exit")
        exit()
```

**Comment by WillAyd** [comment]
Haven't looked deeply but if it helps Ctrl+C throws a KeyboardInterrupt which is a subclass of Exception. So it's quite likely that you have an `except Exception` somewhere which catches the user pressing this combination and somehow then prevents the program from exiting

**Comment by 0golovatyi** [comment]
I did some reading and it seems like there is no universal solution to the problem. On Windows it is `Ctrl+Break`, on Mac it is `Ctrl+\`.
https://stackoverflow.com/questions/1364173/stopping-python-using-ctrlc

**Comment by WillAyd** [comment]
Yea I think that article may be glossing over the fact that the various combinations send different signals to the running process though. Ctrl+C sends a SIGINT which in the Python side would toss a `KeyboardInterrupt`. Realistically the program should handle this gracefully and exit, but within Python that `KeyboardInterrupt` falls under the hierarchy of the base `Exception` (which is why `except` and `except Exception` are strongly discouraged)

It's likely that any of these `except Exception` clauses are the culprint:

```sh
$ grep "except Exception" -r .
./tabpy/tabpy_server/handlers/endpoint_handler.py:        except Exception as e:
./tabpy/tabpy_server/handlers/endpoint_handler.py:            except Exception as e:
./tabpy/tabpy_server/handlers/endpoint_handler.py:                except Exception as e:
./tabpy/tabpy_server/handlers/endpoint_handler.py:        except Exception as e:
./tabpy/tabpy_server/handlers/endpoints_handler.py:            except Exception as ex:
./tabpy/tabpy_server/handlers/endpoints_handler.py:        except Exception as e:
./tabpy/tabpy_server/handlers/evaluation_plane_handler.py:        except Exception as e:
./tabpy/tabpy_server/handlers/management_handler.py:            except Exception as e:
./tabpy/tabpy_server/handlers/query_plane_handler.py:        except Exception as e:
./tabpy/tabpy_server/handlers/query_plane_handler.py:        except Exception as e:
./tabpy/tabpy_server/management/state.py:        except Exception as e:
./tabpy/tabpy_server/management/state.py:        except Exception as e:
./tabpy/tabpy_server/management/state.py:            except Exception as e:
./tabpy/tabpy_server/management/state.py:        except Exception as e:
./tabpy/tabpy_server/management/state.py:        except Exception as e:
./tabpy/tabpy_server/management/state.py:        except Exception as e:
./tabpy/tabpy_server/management/state.py:        except Exception as e:
./tabpy/tabpy_server/management/state.py:        except Exception as e:
./tabpy/tabpy_server/management/state.py:        except Exception as e:
./tabpy/tabpy_server/management/state.py:        except Exception as e:
./tabpy/tabpy_server/management/state.py:        except Exception as e:
./tabpy/tabpy_server/management/state.py:        except Exception as e:
./tabpy/tabpy_server/management/state.py:        except Exception:
./tabpy/tabpy_server/management/state.py:        except Exception:
./tabpy/tabpy_server/management/state.py:        except Exception as e:
./tabpy/tabpy_server/psws/callbacks.py:        except Exception as e:
./tabpy/tabpy_server/psws/callbacks.py:    except Exception as e:
./tabpy/tabpy_server/psws/python_service.py:        except Exception as e:
./tabpy/tabpy_server/psws/python_service.py:        except Exception as e:
./tabpy/tabpy_server/psws/python_service.py:        except Exception as e:
./tabpy/tabpy_server/psws/python_service.py:        except Exception as e:
./tabpy/tabpy_tools/custom_query_object.py:        except Exception as e:
./tabpy/tabpy_tools/custom_query_object.py:        except Exception as e:
./tabpy/tabpy_tools/schema.py:    except Exception as e:
```

Are you seeing anything in the logs when this hangs?

**Comment by 0golovatyi** [comment]
@WillAyd I set breakpoints for all of those lines and neither is activated when Ctrl+C pressed.

**Comment by WillAyd** [comment]
Hmm that's unfortunate. And nothing generally shows in the logs while this is hanging right?

**Comment by 0golovatyi** [comment]
Yes, that is the mystery I am trying to investigate - nothing seems to trigger when I press Ctrl+C on my machine.

**Comment by WillAyd** [comment]
Ah OK. Might just be specific to how the tornado event loop handles things. Maybe this would work?

https://stackoverflow.com/a/17105220/621736

**Comment by 0golovatyi** [comment]
Yes, that solution works, thank you for finding it! It seems wastefull to pull `try_exit` method every so often though... :(
I'll prepare PR later. Thanks again!

**Comment by 0golovatyi** [comment]
Fixed in v 0.8.9 - https://github.com/tableau/TabPy/blob/master/docs/server-config.md#configuration-file-content
